From e464c8d67ca70032dbe1a216ba7bee5ed35e1ac0 Mon Sep 17 00:00:00 2001
From: Nitesh Lal <NiteshNarayanLal@freescale.com>
Date: Mon, 10 Mar 2014 16:57:50 +0530
Subject: [PATCH 1258/1336] crypto: caam - Fix crash in caamalg_qi module_init

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Added NULL check for private structure in caamalg_qi module. It is NULL

when the caam driver fails to properly initialize (e.g. RNG4 init

failed), then the module should gracefully do a failed initialization.

Signed-off-by: Nitesh Lal <NiteshNarayanLal@freescale.com>
Change-Id: Ief8596b9fcc5e16384e4b71ae70ec19b959ea201
Reviewed-on: http://git.am.freescale.net:8181/9578
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Yashpal Dutta <yashpal.dutta@freescale.com>
Reviewed-by: Jose Rivera <German.Rivera@freescale.com>
---
 drivers/crypto/caam/caamalg_qi.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/crypto/caam/caamalg_qi.c b/drivers/crypto/caam/caamalg_qi.c
index 7a476e4..e43a286 100644
--- a/drivers/crypto/caam/caamalg_qi.c
+++ b/drivers/crypto/caam/caamalg_qi.c
@@ -1558,6 +1558,12 @@ static int __init caam_qi_algapi_init(void)
 
 	ctrldev = &pdev->dev;
 	priv = dev_get_drvdata(ctrldev);
+	/*
+	* If priv is NULL, it's probably because the caam driver wasn't
+	* properly initialized (e.g. RNG4 init failed). Thus, bail out here.
+	*/
+	if (!priv)
+		return -ENODEV;
 	of_node_put(dev_node);
 
 	INIT_LIST_HEAD(&alg_list);
-- 
2.0.2

